File: library/src/main/java/me/proxer/library/api/UcpSettingConstraintAdapter.java
package me.proxer.library.api;

import com.squareup.moshi.FromJson;
import com.squareup.moshi.Json;
import com.squareup.moshi.ToJson;
import me.proxer.library.enums.UcpSettingConstraint;

import java.lang.reflect.Field;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * @author Ruben Gees
 */
class UcpSettingConstraintAdapter {

    private final Map<Integer, UcpSettingConstraint> nameToValue;
    private final Map<UcpSettingConstraint, Integer> valueToName;

    UcpSettingConstraintAdapter() {
        final Field[] fields = UcpSettingConstraint.class.getFields();

        nameToValue = new LinkedHashMap<>(fields.length);
        valueToName = new LinkedHashMap<>(fields.length);

        for (final Field field : fields) {
            final Json annotation = field.getAnnotation(Json.class);

            if (annotation != null) {
                final int name = Integer.parseInt(annotation.name());
                final UcpSettingConstraint value = UcpSettingConstraint.valueOf(field.getName());

                nameToValue.put(name, value);
                valueToName.put(value, name);
            }
        }
    }

    @ToJson
    int toJson(final UcpSettingConstraint constraint) {
        return valueToName.get(constraint);
    }

    @FromJson
    UcpSettingConstraint fromJson(final int value) {
        return nameToValue.get(value);
    }
}

File: library/src/main/kotlin/me/proxer/library/api/UcpSettingConstraintAdapter.kt
package me.proxer.library.api

import com.squareup.moshi.FromJson
import com.squareup.moshi.Json
import com.squareup.moshi.ToJson
import me.proxer.library.enums.UcpSettingConstraint
import java.lang.reflect.Field

/**
 * @author Ruben Gees
 */
internal class UcpSettingConstraintAdapter {

    private val nameToValue: Map<Int, UcpSettingConstraint>
    private val valueToName: Map<UcpSettingConstraint, Int>

    init {
        val nameValuePairs = UcpSettingConstraint::class.java.fields.map {
            it.constraintName to it.constraintValue
        }

        nameToValue = nameValuePairs.toMap()
        valueToName = nameValuePairs.map { (name, value) -> value to name }.toMap()
    }

    @ToJson
    fun toJson(constraint: UcpSettingConstraint): Int {
        return valueToName.getValue(constraint)
    }

    @FromJson
    fun fromJson(value: Int): UcpSettingConstraint {
        return nameToValue.getValue(value)
    }

    private inline val Field.constraintName
        get() = Integer.parseInt(this.getAnnotation(Json::class.java).name)

    private inline val Field.constraintValue
        get() = enumValueOf<UcpSettingConstraint>(name)
}
